src/project/git/GitCommandGitProject.ts

```diff
@@ -294,26 +319,73 @@ export class GitCommandGitProject extends NodeFsLocalProject implements GitProje
+               directoryManager: DirectoryManager,
+               secondTry: boolean = false): Promise<GitCommandGitProject> {
     return directoryManager.directoryFor(id.owner, id.repo, id.sha, opts)
         .then(cloneDirectoryInfo => {
-                switch (cloneDirectoryInfo.type) {
-                    case "parent-directory" :
-                        const repoDir = `${cloneDirectoryInfo.path}/${id.repo}`;
-                        const command = (id.sha === "master") ?
-                            `git clone --depth 1 ${id.cloneUrl(credentials)}` :
-                            `git clone ${id.cloneUrl(credentials)}; cd ${id.repo};git checkout ${id.sha}`;
-
-                        logger.info(`Cloning repo '${id.url}' to '${cloneDirectoryInfo.path}'`);
-                        return exec(command, { cwd: cloneDirectoryInfo.path })
-                            .then(_ => {
-                                logger.debug(`Clone succeeded with URL '${id.url}'`);
-                                // fs.chmodSync(repoDir, "0777");
-                                return NodeFsLocalProject.fromExistingDirectory(id, repoDir);
+            switch (cloneDirectoryInfo.type) {
+                case "empty-directory" :
+                    return cloneInto(credentials, cloneDirectoryInfo, id);
+                case "existing-directory" :
+                    const repoDir = cloneDirectoryInfo.path;
+                    return checkout(repoDir, id.sha)
+                        .then(() => clean(repoDir))
+                        .then(() => {
+                            return GitCommandGitProject.fromBaseDir(id,
+                                repoDir, credentials, cloneDirectoryInfo.release,
+                                cloneDirectoryInfo.provenance + "\nRe-using existing clone");
+                        }, error => {
+                            return cloneDirectoryInfo.invalidate().then(() => {
+                                if (secondTry) {
+                                    throw error;
+                                } else {
+                                    return clone(credentials, id, opts, directoryManager, true);
+                                }
                             });
-                    case "actual-directory" :
-                        throw new Error("actual-directory clone directory type not yet supported");
-                }
-            },
-        );
+                        });
+                default:
+                    throw new Error("What is this type: " + cloneDirectoryInfo.type);
+            }
+        });
+}
+
+function cloneInto(credentials: ProjectOperationCredentials,
+                   targetDirectoryInfo: CloneDirectoryInfo, id: RemoteRepoRef) {
+    const repoDir = targetDirectoryInfo.path;
+    const command = (id.sha === "master" && targetDirectoryInfo.transient) ?
+        runIn(".", `git clone --depth 1 ${id.cloneUrl(credentials)} ${repoDir}`) :
```

Comment:
this --depth 1 optimization now runs only for transient directories (temporary ones). If this is our cached clone we want a full one, please.